### ISSUE TITLE
[BUG] Numerical instability in loss curve over epochs when training on A100 with TF32 precision

### ISSUE THREAD

**Issue by rnyak** [NVIDIA-Merlin/models#857]
### Bug description
We have been reported by a user that when they train a TT model on P4 instance on AWS they observe the unstable loss curve as below. And this is mostly because of `TF32` precision on Ampere GPU. The blue line below is `TF32` and green is `FP32` on the same machine.

![Screen Shot 2022-10-28 at 9 01 38 AM](https://user-images.githubusercontent.com/16246900/199797549-9491dde2-254c-47c3-84ab-a9f47ba9388e.png)

Currently this is being debugged and we run experiments  to diagnose the issue.

**Comment by viswa-nvidia** [comment]
@rnyak , please triage this bug.

**Comment by gabrielspmoreira** [comment]
@rnyak As you know, I performed a number of experiments with Two-Tower models (V1) on both V100 and A100 trying to reproduce the issue with a public dataset. I was able to find some situations where the loss became unstable, but not much due to the usage of Ampere (A100 x V100) or to the usage of the new TensorFloat32 (TF32) support in Ampere compared to FP32 ops, but to other factors like high Learning Rate.

One thing to notice is that the `loss` metric reported for each step by Keras is the cumulative average of the loss within the epoch. The plot provided above is the default logging with `TensorboardCallback`, which logs the (avg per-epoch) `loss` in the end of each epoch. To be able to better investigate the loss instability, I introduce in PR #862 the `loss_batch` metric, which reports the actual loss per-batch and not the cumulative per-epoch average `loss`. 

In the following plot for A100 runs, we have the `loss_batch` on the left for `FP32`, in the middle for TF32, and on the right the cumulative per-epoch avg. `loss` for both. 
P.s. Those peaks at the avg `loss` in the beginning of each epoch happen because the cumulative per-epoch average is more sensitive to the peaks (high values) in the first training steps of each epoch.
We can notice that with `TP32` the peaks are slightly higher than with `FP32` but the general trend of avg loss is the same.
![image](https://user-images.githubusercontent.com/504752/200353559-6f6ee68b-8eb7-499a-8832-d214318714ad.png).

The most important factor we could find related to loss instability was the usage of high LR. For example, below we plot in red the `loss_batch` with LR `1e-3` and in purple with LR `1e-4` (both with Adam optimizer on A100 with TP32). With the smaller `LR` we don't observe such instability.
![image](https://user-images.githubusercontent.com/504752/200354791-cf1e9a00-9000-41be-bb9d-fc4c624f4dd8.png)

Could the reported of the issue try using a smaller constant LR or a learning rate schedule (e.g. `tf.keras.optimizers.schedules.ExponentialDecay`) to try and stabilize the loss. If the issue persist, could the reporter try to log in TensorBoardCallback or WandbCallBack (W&B) the metrics per batch, including the new `loss_batch` metric.

**Comment by EderSantana** [comment]
Oi @gabrielspmoreira !! I'm the original author of the graphs on the top of the ticket. So just to clarify those two learning curves were generated using the same set of hyperparameters. Including the same LR. The only thing that I changed was TF32 and FP32 format. So I'm still not sure if different learning rates would explain everything.

That said, I also used AWS P4s to get those GPUs. I don't know if the way AWS provides those instances would have any extra configurations affecting the results. But I wouldn't expect so... Do you have any previous experience with nvidia drivers or anything like that affecting numerical stability?

**Comment by gabrielspmoreira** [comment]
oi @EderSantana ! Nice to meet you!
I understand you have observed the difference by just enabling/disabling TF32.
I used in the experiments both `TwoTowerModel` and your custom implementation of it (with shared embeddings between towers), Adam and LR 1e-3. But unfortunately I could not reproduce the issue with a public dataset (TenRec).

Would you be able to provide a reproducer (exact code, hparams and data sample), so that we isolate if the issue is within Merlin Models, TF or Nvidia drivers?

May I ask if the issue also happens with the `MatrixFactorizationModel`? As that one uses only user and item id and a dot product operation, that could be a good test to better isolate whether the issue comes with additional features and dense layers.
In addition, have you tried the `InputBlockV2` and `TwoTowerModelV2` example I think @rnyak has provided you? Porting from the "V1" of those classes to V2 should be straightforward and then you could check if the loss instability issue remains with the latest version.

**Comment by EderSantana** [comment]
It would be hard for me to share more due to data privacy issues and I talked to Ronay... 
But I'll give InputBlockV2 a try as soon as possible let you know how it goes.
Thanks a lot for checking it out and clarifying to me here!

**Comment by sararb** [comment]
Closing this ticket as we couldn't reproduce the loss instability with synthetic data. Please re-open a new issue if the instability is observed with the `InputBlockV2` as well.
